README.md
# poyo

A lightweight YAML Parser for Python. 🐓

**poyo** does not allow deserialization of arbitrary Python objects. Supported
types are `str`, `bool`, `int`, `float`, `NoneType` as well as `dict` and
`list` values.

⚠️ Please note that poyo supports only a chosen subset of the YAML format
that is required to parse [cookiecutter user configuration
files][cookiecutterrc]. poyo does not have support for serializing into YAML
and is not compatible with JSON.

[cookiecutterrc]: https://cookiecutter.readthedocs.io/en/latest/advanced/user_config.html

## Installation

poyo is available on [PyPI][PyPI] for Python versions 2.7 and newer and can
be installed with [pip][pip]:

```text
pip install poyo
```

[PyPI]: https://pypi.org/project/poyo/
[pip]: https://pypi.org/project/pip/

This package does not have any additional requirements. 📦

## Usage

poyo comes with a ``parse_string()`` function, to load utf-8 encoded string
data into a Python dict.

```python
import codecs
import logging

from poyo import parse_string, PoyoException

logging.basicConfig(level=logging.DEBUG)

with codecs.open("tests/foobar.yml", encoding="utf-8") as ymlfile:
    ymlstring = ymlfile.read()

try:
    config = parse_string(ymlstring)
except PoyoException as exc:
    logging.error(exc)
else:
    logging.debug(config)
```

## Example

### Input YAML string

```yaml
---
default_context: # foobar
    greeting: こんにちは
    email: "raphael@hackebrot.de"
    docs: true

    gui: FALSE
    123: 456.789
    # comment
    # allthethings
    'some:int': 1000000
    foo: "hallo #welt" #Inline comment :)
    longtext: >
        This is a multiline string.
        It can contain all manners of characters.

        Single line breaks are ignored,
        but blank linkes cause line breaks.
    trueish: Falseeeeeee
    blog   : raphael.codes
    relative-root: /          # web app root path (default: '')
    lektor: 0.0.0.0:5000      # local build
    doc_tools:
        # docs or didn't happen
        -    mkdocs
        - 'sphinx'

        - null
    # 今日は
zZz: True
NullValue: Null

# Block
# Comment

Hello World:
    # See you at EuroPython
    null: This is madness   # yo
    gh: https://github.com/{0}.git
"Yay #python": Cool!
```

### Output Python dict

```python
{
    u"default_context": {
        u"greeting": u"こんにちは",
        u"email": u"raphael@hackebrot.de",
        u"docs": True,
        u"gui": False,
        u"lektor": "0.0.0.0:5000",
        u"relative-root": "/",
        123: 456.789,
        u"some:int": 1000000,
        u"foo": u"hallo #welt",
        u"longtext": (
            u"This is a multiline string. It can contain all "
            u"manners of characters.\nSingle line breaks are "
            u"ignored, but blank linkes cause line breaks.\n"
        ),
        u"trueish": u"Falseeeeeee",
        u"blog": u"raphael.codes",
        u"doc_tools": [u"mkdocs", u"sphinx", None],
    },
    u"zZz": True,
    u"NullValue": None,
    u"Hello World": {
        None: u"This is madness",
        u"gh": u"https://github.com/{0}.git",
    },
    u"Yay #python": u"Cool!",
}
```

## Logging

poyo follows the recommendations for [logging in a library][logging], which
means it does not configure logging itself. Its root logger is named ``poyo``
and the names of all its children loggers track the package/module hierarchy.
poyo logs to a ``NullHandler`` and solely on ``DEBUG`` level.

If your application configures logging and allows debug messages to be shown,
you will see logging when using poyo. The log messages indicate which parser
method is used for a given string as the parser deseralizes the config. You
can remove all logging from poyo in your application by setting the log level
of the ``poyo`` logger to a value higher than ``DEBUG``.

[logging]: https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library

### Disable Logging

```python
import logging

logging.getLogger("poyo").setLevel(logging.WARNING)
```

### Example Debug Logging Config

```python
import logging
from poyo import parse_string

logging.basicConfig(level=logging.DEBUG)

CONFIG = """
---
default_context: # foobar
    greeting: こんにちは
    gui: FALSE
    doc_tools:
        # docs or didn't happen
        -    mkdocs
        - 'sphinx'
    123: 456.789
"""

parse_string(CONFIG)
```

### Example Debug Logging Messages

```text
DEBUG:poyo.parser:parse_blankline <- \n
DEBUG:poyo.parser:parse_blankline -> IGNORED
DEBUG:poyo.parser:parse_dashes <- ---\n
DEBUG:poyo.parser:parse_dashes -> IGNORED
DEBUG:poyo.parser:parse_section <- default_context: # foobar\n
DEBUG:poyo.parser:parse_str <- default_context
DEBUG:poyo.parser:parse_str -> default_context
DEBUG:poyo.parser:parse_section -> <Section name: default_context>
DEBUG:poyo.parser:parse_simple <-     greeting: \u3053\u3093\u306b\u3061\u306f\n
DEBUG:poyo.parser:parse_str <- greeting
DEBUG:poyo.parser:parse_str -> greeting
DEBUG:poyo.parser:parse_str <- \u3053\u3093\u306b\u3061\u306f
DEBUG:poyo.parser:parse_str -> \u3053\u3093\u306b\u3061\u306f
DEBUG:poyo.parser:parse_simple -> <Simple name: greeting, value: \u3053\u3093\u306b\u3061\u306f>
DEBUG:poyo.parser:parse_simple <-     gui: FALSE\n
DEBUG:poyo.parser:parse_str <- gui
DEBUG:poyo.parser:parse_str -> gui
DEBUG:poyo.parser:parse_false <- FALSE
DEBUG:poyo.parser:parse_false -> False
DEBUG:poyo.parser:parse_simple -> <Simple name: gui, value: False>
DEBUG:poyo.parser:parse_list <-     doc_tools:\n        # docs or didn't happen\n        -    mkdocs\n        - 'sphinx'\n
DEBUG:poyo.parser:parse_str <- mkdocs
DEBUG:poyo.parser:parse_str -> mkdocs
DEBUG:poyo.parser:parse_str <- 'sphinx'
DEBUG:poyo.parser:parse_str -> sphinx
DEBUG:poyo.parser:parse_str <- doc_tools
DEBUG:poyo.parser:parse_str -> doc_tools
DEBUG:poyo.parser:parse_list -> <Simple name: doc_tools, value: ['mkdocs', 'sphinx']>
DEBUG:poyo.parser:parse_simple <-     123: 456.789\n
DEBUG:poyo.parser:parse_int <- 123
DEBUG:poyo.parser:parse_int -> 123
DEBUG:poyo.parser:parse_float <- 456.789
DEBUG:poyo.parser:parse_float -> 456.789
DEBUG:poyo.parser:parse_simple -> <Simple name: 123, value: 456.789>
DEBUG:poyo.parser:parse_simple <-     docs: true\n
DEBUG:poyo.parser:parse_str <- docs
DEBUG:poyo.parser:parse_str -> docs
DEBUG:poyo.parser:parse_true <- true
DEBUG:poyo.parser:parse_true -> True
DEBUG:poyo.parser:parse_simple -> <Simple name: docs, value: True>
```

## About this project

We created this project to work around installation issues with a
[cookiecutter][cookiecutter] version that depended on existing YAML parsers
for Python. For more information please check out this [GitHub issue][issue].

[issue]: https://github.com/cookiecutter/cookiecutter/pull/621

## Community

Would you like to contribute to **poyo**? You're awesome! 😃

Please check out the [good first issue][good first issue] label for tasks,
that are good candidates for your first contribution to poyo. Your
contributions are greatly appreciated! Every little bit helps and credit will
always be given.

Everyone interacting in the poyo project's codebases, issue trackers, chat
rooms, and mailing lists is expected to follow the [PyPI Code of
Conduct][code of conduct].

Join the poyo [community][community]! 🌍🌏🌎

[code of conduct]: https://www.pypa.io/en/latest/code-of-conduct/
[community]: https://github.com/hackebrot/poyo/blob/master/COMMUNITY.md
[good first issue]: https://github.com/hackebrot/poyo/labels/good%20first%20issue

## License

Distributed under the terms of the [MIT][MIT] license, poyo is free and open source
software.

[MIT]: https://github.com/hackebrot/poyo/blob/master/LICENSE

[cookiecutter]: https://github.com/cookiecutter/cookiecutter


CHANGES.md
# 0.6.0

### Improvements

* Add a failing test for list objects, thanks to [@lukaselmer][@lukaselmer]
  (#19)

[@lukaselmer]: https://github.com/lukaselmer

# 0.5.0

### Features

* Add support for multiline strings, thanks to [@maebert][@maebert] (#22)

    ```yaml
    longtext: >
        This is a multiline string.
        It can contain all manners of characters.

        Single line breaks are ignored,
        but blank linkes cause line breaks.
    ```

### Bugfixes

* Fix link to poyo on PyPI, thanks to [@mrshu][@mrshu] (#20)
* Trim trailing whitespace in variable, thanks to [@hackebrot][@hackebrot] (#25)

### Improvements

* Update project to use src directory, thanks to [@hackebrot][@hackebrot] (#24)
* Convert rst files to md, thanks to [@Um9i][@Um9i] (#28)
* Reformat code base with black, thanks to [@Um9i][@Um9i] (#29)

[@maebert]: https://github.com/maebert
[@Um9i]: https://github.com/Um9i
[@hackebrot]: https://github.com/hackebrot
[@mrshu]: https://github.com/mrshu

# 0.4.2

### Bugfixes

* Resolve a bug with list items and comments, thanks to
  [@ishanarora][@ishanarora] (#18)

[@ishanarora]: https://github.com/ishanarora

# 0.4.1

### Bugfixes

* Update regex patterns to allow for no newline at the end of the given string,
  thanks to [@mikeckennedy][@mikeckennedy] and [@hackebrot][@hackebrot]
  (#13, #14)

[@mikeckennedy]: https://github.com/mikeckennedy

# 0.4.0

### Features

* Add support for block comments in sections, thanks to [@jakubka][@jakubka]
  and [@hackebrot][@hackebrot] (#7)

    ```yaml
    default_context: # foobar
        greeting: こんにちは
        # comment
        # allthethings
        docs: true

        123: 456.789
    ```

### Improvements

* Set up ``poyo`` logger with NullHandler to log DEBUG messages when parsing,
  thanks to [@hackebrot][@hackebrot]

    ```text
    DEBUG:poyo.parser:parse_simple <-     123: 456.789
    DEBUG:poyo.parser:parse_int <- 123
    DEBUG:poyo.parser:parse_int -> 123
    DEBUG:poyo.parser:parse_float <- 456.789
    DEBUG:poyo.parser:parse_float -> 456.789
    DEBUG:poyo.parser:parse_simple -> <Simple name: 123, value: 456.789>
    ```

### Bugfixes

* Fix an issue around section names if the line contained more than one colon
  symbol, thanks to [@gvalkov][@gvalkov] and [@hackebrot][@hackebrot] (#9)
* Fix an issue that caused partial matches to raise an error, thanks to
  [@gvalkov][@gvalkov] and [@hackebrot][@hackebrot] (#9)

[@gvalkov]: https://github.com/gvalkov
[@hackebrot]: https://github.com/hackebrot
[@jakubka]: https://github.com/jakubka


# 0.3.0

### Features

* Add support for blank lines and comment lines in lists, thanks to
  [@eykd][@eykd] and [@hackebrot][@hackebrot] (#5)

    ```yaml
    doc_tools:
        # docs or didn't happen
        -    mkdocs
        - 'sphinx'

        - null
    ```

### Improvements

* Add tests for patterns, thanks to [@eykd][@eykd] and [@hackebrot][@hackebrot]
  (#5)

### Bugfixes

* Solve an issue with ``~`` character not being recognized as ``None``

[@eykd]: https://github.com/eykd
[@hackebrot]: https://github.com/hackebrot


# 0.2.0

### Features

* Add support for list values

    ```yaml
    doc_tools:
        - mkdocs
        - 'sphinx'
        - null
    ```

* Expose ``PoyoException`` in API

    ```python
    from poyo import PoyoException
    ```

### Bugfixes

* Ignore dashes in lines

    ```yaml
    ---
    default_context:
        foo: "hallo #welt" #Inline comment :)
        docs: true
    ```

# 0.1.0

First release on PyPI.

### Features

* ``parse_string()`` to load a YAML string as a Python dict


src/poyo/parser.py
# -*- coding: utf-8 -*-

import functools
import logging

from ._nodes import Root, Section, Simple

from .exceptions import (
    NoMatchException, NoParentException, NoTypeException, IgnoredMatchException
)

from .patterns import (
    COMMENT, BLANK_LINE, DASHES, LIST, SIMPLE, SECTION,
    LIST_ITEM, NULL, TRUE, FALSE, FLOAT, INT, STR, MULTILINE_STR
)

logger = logging.getLogger(__name__)


def log_callback(wrapped_function):
    """Decorator that produces DEBUG level log messages before and after
    calling a parser method.

    If a callback raises an IgnoredMatchException the log will show 'IGNORED'
    instead to indicate that the parser will not create any objects from
    the matched string.

    Example:
        DEBUG:poyo.parser:parse_simple <-     123: 456.789
        DEBUG:poyo.parser:parse_int <- 123
        DEBUG:poyo.parser:parse_int -> 123
        DEBUG:poyo.parser:parse_float <- 456.789
        DEBUG:poyo.parser:parse_float -> 456.789
        DEBUG:poyo.parser:parse_simple -> <Simple name: 123, value: 456.789>
    """

    def debug_log(message):
        """Helper to log an escaped version of the given message to DEBUG"""
        logger.debug(message.encode('unicode_escape').decode())

    @functools.wraps(wrapped_function)
    def _wrapper(parser, match, **kwargs):
        func_name = wrapped_function.__name__

        debug_log(u'{func_name} <- {matched_string}'.format(
            func_name=func_name,
            matched_string=match.group(),
        ))

        try:
            result = wrapped_function(parser, match, **kwargs)
        except IgnoredMatchException:
            debug_log(u'{func_name} -> IGNORED'.format(func_name=func_name))
            raise

        debug_log(u'{func_name} -> {result}'.format(
            func_name=func_name,
            result=result,
        ))

        return result
    return _wrapper


class _Parser(object):
    def __init__(self, source):
        self.pos = 0
        self.source = source
        self.max_pos = len(self.source)

        self.root = Root()
        self.seen = [self.root]

        self.rules = (
            (COMMENT, self.parse_comment),
            (BLANK_LINE, self.parse_blankline),
            (DASHES, self.parse_dashes),
            (LIST, self.parse_list),
            (MULTILINE_STR, self.parse_multiline_str),
            (SIMPLE, self.parse_simple),
            (SECTION, self.parse_section),
        )

        self.tag_rules = (
            (NULL, self.parse_null),
            (TRUE, self.parse_true),
            (FALSE, self.parse_false),
            (INT, self.parse_int),
            (FLOAT, self.parse_float),
            (STR, self.parse_str),
        )

    def find_at_level(self, level):
        for candidate in reversed(self.seen):
            if candidate.level < level:
                return candidate
        raise NoParentException(
            'Unable to find element at level {}'.format(level)
        )

    def read_from_tag(self, string):
        for pattern, callback in self.tag_rules:
            match = pattern.match(string)

            if not match:
                # Pattern does not match, try the next one
                continue

            if match.group() != string:
                # Unable to match the full string, try the next pattern
                continue

            return callback(match)

        error_message = 'Unable to determine type for "{}"'
        raise NoTypeException(error_message.format(string))

    @log_callback
    def parse_comment(self, match):
        """Ignore line comments."""
        raise IgnoredMatchException

    @log_callback
    def parse_blankline(self, match):
        """Ignore blank lines."""
        raise IgnoredMatchException

    @log_callback
    def parse_dashes(self, match):
        """Ignore lines that contain three dash symbols."""
        raise IgnoredMatchException

    @log_callback
    def parse_list(self, match):
        groups = match.groupdict()
        level = len(groups['indent'])
        parent = self.find_at_level(level)

        item_matches = LIST_ITEM.findall(groups['items'])

        list_items = [
            self.read_from_tag(value)
            for value in item_matches
        ]

        variable = self.read_from_tag(groups['variable'])
        return Simple(variable, level, list_items, parent=parent)

    @log_callback
    def parse_simple(self, match):
        groups = match.groupdict()

        level = len(groups['indent'])
        parent = self.find_at_level(level)

        variable = self.read_from_tag(groups['variable'])
        value = self.read_from_tag(groups['value'])

        return Simple(variable, level, value, parent=parent)

    @log_callback
    def parse_section(self, match):
        groups = match.groupdict()

        level = len(groups['indent'])
        parent = self.find_at_level(level)

        return Section(
            self.read_from_tag(groups['variable']),
            level,
            parent=parent
        )

    @log_callback
    def parse_null(self, match):
        return None

    @log_callback
    def parse_true(self, match):
        return True

    @log_callback
    def parse_false(self, match):
        return False

    @log_callback
    def parse_int(self, match):
        return int(match.group())

    @log_callback
    def parse_float(self, match):
        return float(match.group())

    @log_callback
    def parse_str(self, match):
        quotes = match.group('quotes')
        return match.group().strip(quotes)

    def join_lines(self, lines, keep_newlines=False):
        result = ""
        concat = "\n" if keep_newlines else " "
        for line in lines:
            if line:
                result += line + concat
            else:
                result += "\n"
        return result.rstrip(" ").replace(" \n", "\n")

    @log_callback
    def parse_multiline_str(self, match):
        groups = match.groupdict()
        keep_newlines = groups['blockstyle'] == "|"
        chomp = groups['chomping']
        level = len(groups['indent'])
        parent = self.find_at_level(level)
        variable = self.read_from_tag(groups['variable'])
        lines = groups['lines'].splitlines()
        if not lines:
            return Simple(variable, level, "", parent=parent)

        first_indent = groups['forceindent']
        if first_indent:
            first_indent = int(first_indent)
        else:
            first_indent = len(lines[0]) - len(lines[0].lstrip())
        value = self.join_lines(
            [l[first_indent:] for l in lines],
            keep_newlines
        )
        if not chomp:
            value = value.rstrip("\n") + "\n"
        elif chomp == "-":
            value = value.rstrip("\n")
        return Simple(variable, level, value.rstrip(""), parent=parent)

    def find_match(self):
        """Try to find a pattern that matches the source and calll a parser
        method to create Python objects.

        A callback that raises an IgnoredMatchException indicates that the
        given string data is ignored by the parser and no objects are created.

        If none of the pattern match a NoMatchException is raised.
        """
        for pattern, callback in self.rules:
            match = pattern.match(self.source, pos=self.pos)

            if not match:
                continue

            try:
                node = callback(match)
            except IgnoredMatchException:
                pass
            else:
                self.seen.append(node)

            return match

        raise NoMatchException(
            'None of the known patterns match for {}'
            ''.format(self.source[self.pos:])
        )

    def __call__(self):
        """Parse the given string data and sequentually update the current
        cursor position until the end is reached.

        Return the Root object if successful.
        """
        while self.pos < self.max_pos:
            match = self.find_match()
            self.pos = match.end()
        return self.root()


def parse_string(string):
    parser = _Parser(string)
    return parser()
